src/tls/certfile.c

```diff
@@ -0,0 +1,301 @@
+    goto out;
+
+  /* We need to take the mutex here two prevent (at least) two problems:
+   *
+   *  - two connections starting at the same time could get in a fight
+   *    about which one is responsible for writing the contents of the
+   *    certificate to the file in case they both fstat() before either
+   *    of them writes.  This is not very serious, but it is undesired.
+   *
+   *  - a connection starting just as the last connection is exiting on
+   *    another thread could open the certificate file, not yet
+   *    acquiring its lock, just before the file is unlinked by the
+   *    exiting thread.  We would then successfully acquire a lock on
+   *    the no-longer-linked file.  This is very serious.
+   *
+   * Mutual exclusion solves both of those issues.
+   */
+  pthread_mutex_lock (&certfile_mutex);
+
+  /* We have two cases here: the case where the file already exists, and
+   * the case where we must create the file.
+   *
+   * We attempt a separate open() for each of these two cases because we
+   * need to detect which situation we are in because the error handling
+   * is different for each case: if something goes wrong while we're
+   * attempting to create the file, then we need to make sure we unlink
+   * it again in case of an error.  Otherwise, we need to leave it
+   * alone.
+   */
+  fd = openat (dirfd, fingerprint.str, O_RDWR, 0666);
+  if (fd == -1)
+    {
+      if (errno != ENOENT)
+        {
+          warn ("Couldn't open existing fingerprint file %s", fingerprint.str);
+          goto out;
+        }
+
+      /* ENOENT case: The file didn't exist.  Create it. */
+      fd = openat (dirfd, fingerprint.str, O_CREAT | O_EXCL | O_RDWR, 0666);
+      if (fd == -1)
+        {
+          /* We're doing this all while holding a lock, so any error at
+           * all at this point (including the file springing into
+           * existence since open() failed with ENOENT) is unexpected.
+           */
+          warn ("Failed to create fingerprint file %s", fingerprint.str);
+          goto out;
+        }
+
+      /* We've successfully created the file.  Any failure should result
+       * in an unlink.
+       */
+      unlink_on_error = fingerprint.str;
+
+      if (pwrite (fd, peer_certificate.data, peer_certificate.size, 0) != peer_certificate.size)
+        {
+          warn ("Couldn't write content to certificate file %s", fingerprint.str);
+          goto out;
+        }
+    }
+
+  /* At this point, we have a valid fd and a file with content in it.
+   *
+   * Write locks are only ever held while also holding the mutex, so if
+   * we fail to acquire a read lock, something has gone seriously wrong.
+   */
+  if (flock (fd, LOCK_SH | LOCK_NB) != 0)
+    {
+      warn ("Couldn't acquire read lock on certificate file %s", fingerprint.str);
+      goto out;
+    }
+
+  /* success */
+  result = fd;
+  unlink_on_error = NULL;
+  fd = -1;
+
+out:
+  (gnutls_free) (peer_certificate.data);
```

Comment:
The parens are because `gnutls_free()` is redefined as a evil side-effecting macro (which multiple-evaluates its paramter, including as an lvalue which it sets to NULL).  Using parens circumvents the macro expansion, effectively invoking the original function.

Reply:
I'll add a comment.